### PR TITLE
[OSS] Getting rid of the "should bucket" hash table, just use a list + robustness related fixes

### LINKED ISSUE #258: [OSS] Params without requires_grad can slip out of the bucketing table, and break
## 🐛 Bug
Root bug behind https://github.com/huggingface/transformers/issues/9156, which explains why a "new" parameter was showing up. The hash table is probably slow and not really needed any way, I should have stuck to a list as recommended by @msbaines long ago

## Expected behavior
Should never break 

cc @stas00

### PR DESCRIPTION
**Pull request by blefaudeux**
Properly handle all params, with or without requires_grad, HuggingFace has triggered a bug which was always there. Remove the hash table to store the strategy (a bit risky if the tensors are moved around) and replace with a simple list. Make sure that we update the strategy if a new parameter group is added

# Before submitting
- [x] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
- [x] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/master/CONTRIBUTING.md)?
- [ ] Did you make sure to update the docs?
- [x] Did you write any new necessary tests?

## What does this PR do?
Fixes https://github.com/huggingface/transformers/issues/9156 for the multi-node case, ie: with bucketing involved, tested locally by commenting out the no-bucketing statement.
Fixes #258 
Makes sure that this case is caught by the unit test, I'm really surprised that this was not covered before, but I did check that master fails on this

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Well, yes 🙃 the fp16 issue is not solved though

cc @stas00

### CODE REVIEW

**Comment by stas00** [comment]
Thank you very much for the fix, @blefaudeux!

**Comment by blefaudeux** [comment]
ping reviews, if you don't mind, I would love master to work for HuggingFace

File: tests/nn/data_parallel/test_sharded_ddp.py [hunk 1]
```diff
@@ -44,6 +44,7 @@ def check(broadcast_buffers: bool, grad_accumulation: bool = False) -> None:
         model.register_buffer("test_buffer", torch.ones((1)) * rank)
         model.to(device)
 
+        next(model.parameters()).requires_grad = False
```
Comment (min-xu-ai) [inline tests/nn/data_parallel/test_sharded_ddp.py:47]: add a comment?

**Review by min-xu-ai — APPROVED** [review]
Looks sane to me. Again, it seems the changes are centered around having params that doesn't have require_grads. Perhaps more explicit test cases would help prevent regression?

File: tests/nn/data_parallel/test_sharded_ddp.py (hunk 1 above)
Comment (blefaudeux) [reply tests/nn/data_parallel/test_sharded_ddp.py:47]: Yes, I was about to right that this is the unit test change which makes sure that this is caught in the future. Will do !

**Pull request merged**